### ISSUE TITLE
Create OMR framework for out-of-line object allocation and initialization

### ISSUE THREAD

**Issue by ktbriggs** [eclipse/omr#201]
Client languages such as Java will need to communicate to OMR how to customize heap allocation requests and initialize heap memory that they allocate for objects. 

Proposed framework provides an MM_AllocateInitialization object that client languages can subclass to inform OMR regarding allocation request parameters and to lay out the heap memory allocated to satisfy those requests. 

In addition to defining one or more subclasses of MM_AllocateInitialization, client language is required to define an enumeration of these subclasses in GC_ObjectModel and implement:

```
omrobjectptr_t GC_ObjectModel::initializeAllocation(MM_EnvironmentBase *, void *,
			MM_AllocateInitialization *)
```
which can cast the MM_AllocateInitialization* to the appropriate subclass and complete the initialization of the allocated memory.

**Comment by ktbriggs** [comment]
Note to reviewers: Github shows 2 commits above but d88b820 does not exist on this branch -- it was squashed into 575a293 and force-pushed. Not sure why GitHub is showing it here, but the 575a293  includes all of the changes in d88b820. Please review 575a293 only.

**Comment by ktbriggs** [comment]
Closed via #289.
